src/Domain/BlockVector.hpp

```diff
@@ -0,0 +1,121 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <memory>
+#include <ostream>
+#include <vector>
+
+#include "Domain/BlockId.hpp"
+
+namespace domain {
+/*!
+ * \ingroup ComputationalDomainGroup
+ * \brief A vector that is indexed by `BlockId`.
+ */
+template <class T, class Allocator = std::allocator<T>>
+class BlockVector : private std::vector<T, Allocator> {
```

Comment:
This class does not behave how I would expect a class with a nearly opaque index (as I believe BlockId is intended to be) to behave.  I would expect something that behaves similarly to an associative container rather than a sequence container.  `insert` seems like a strange operation for instance, since it reassociates everything later in the container with a different BlockId.

It seems that this assumes BlockId acts like a pseudo-size_t, when I would expect it to be treated as some sort of more abstract label (like ElementId or the pair used as a mortar id) (although I'm not sure how you would generate such a label for Blocks).

Since we are forced to treat BlockIds as something like size_ts anyway, it is not clear to me that this class benefits us much over just using a normal vector and indexing with `block_id.id` or some such.

Reply:
I was hoping to be able to get right of the `block_id.id` to make it extra difficult to misuse the BlockId. I certainly agree that `insert` does not make any sense in this context! Even things like `push_back` seem a bit questionable but are more reasonable than `insert`. I think what I will do is factor out BlockVector into a separate PR so we can get the rest merged and then decide whether or not we want to have every BlockId related thing go through BlockId only or whether it's possible to extract the size_t out.

Reply:
yes factoring the PR into two sounds good